rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,418 @@
+ANSI-compliant and also present in [spark-sql](https://spark.apache.org/docs/latest/sql-ref-ansi-compliance.html).
+
+```sql
+-- PROPOSED SYNTAX WITH FUNCTION KEYWORD --
+CREATE
+[FUNCTION] INDEX index_name ON table_name [ USING index_type ] ( { column_name | expression } );
+-- Examples --
+CREATE FUNCTION INDEX last_name_idx ON employees (UPPER(last_name)); -- functional index using column stats for UPPER(last_name)
+CREATE FUNCTION INDEX datestr ON hudi_table (DATE_FORMAT(timestamp, '%Y-%m-%d')); -- functional index using column stats for DATE_FORMAT(timestamp, '%Y-%m-%d')
+CREATE INDEX city_id_idx ON hudi_table (city_id); -- usual column stats within MT column_stats partition
+CREATE FUNCTION INDEX hour_of_day ON hudi_table USING BITMAP (DATE_FORMAT(timestamp, '%H')); -- functional index using bitmap for DATE_FORMAT(timestamp, '%H')
+CREATE FUNCTION INDEX income_idx ON employees (salary + (salary*commission_pct)); --  functional index using column stats for given expression
+-- NO CHANGE IN DROP INDEX --
+DROP INDEX last_name_idx;
+```
+
+`index_name` - Required, should be validated by parser. The name will be used for the partition name in MT.
+
+`index_type` - Optional, column\_stats if none provided and there are no functions and expressions in the command. Valid
+options could be BITMAP, COLUMN\_STATS, LUCENE, etc. If index\_type is not provided, and there are functions or
+expressions in the command then a functional index using column stats will be created.
+
+`expression` - simple scalar expressions or sql functions.
+
+#### Functional Index Metadata
+
+For each functional index, store a separate metadata with index details. This should be efficiently loaded. One option
+is to store the below metadata in `hoodie.properties`. This way all index metadata can be loaded with the table config.
+But, it would be better to not overload the table config. Let `hoodie.properties` still hold the list of indexes (MT
+partitions) available, and we propose to create separate `.hudi_index` directory under `basepath/.hoodie`. This has the
+benefit of separation of concern and can be cheaply loaded with the metaclient (one extra file i/o only if there is a
+functional index in the table config). It can store the following metadata for each index.
+
+* Name of the index
+* Expression (e.g., `MONTH(timestamp)`)
+* Data column(s) it's derived from
+* Any other configuration or properties
+
+#### APIs
+
+A new interface, `HoodieFunctionalIndex`, is introduced to represent a functional index and track the relationship
+between the original data column(s) and the transformation.
+
+```java
+interface HoodieFunctionalIndex<S, T> {
+  String indexName;
+  String expression;
+  List<String> columnsInExpression;
+
+  T apply(S sourceValue);
+}
+```
+
+As mentioned in the goals of this RFC, we are not going to build an expression engine. To begin with, we can simply
```

Comment:
would we need any format changes for this later on? can we think through that case?

Reply:
Expression engine is gonna be a higher-level construct, working primarily at the logical layer rather than the physical storage layer. We will store the compiled form of expression and expression engine version itself in `.index_defs`. I don't think more storage changes would be needed.